juju/juju
# Fix refresh when upgrading from pod-spec to sidecar charm

This PR isn't as quite as scary-big as the diffstat shows -- about `+3400 -1700` is due to (generated) changes in `apiserver/facades/schema.json`.

There are a number of changes to make the upgrade work, namely:

* Update SetCharm facade (and state.Application.SetCharm) to update series from "kubernetes" to new base (e.g., focal).
* Don't filter CAAS application change events going to workers by legacy (v1) vs sidecar (v2); filter them in the workers instead so we still receive the events. Application child workers that are v1-specific exit if the charm format changes to v2. (NOTE: I'll remove the filtering watcher in a subsequent PR.)
* When setting up K8s volumes for sidecar applications, ignore volume sources created by pod-spec charms (configMap, hostPath, emptyDir).
* In the caasapplicationprovisioner application worker, wait till it upgrades to a v2 charm before proceeding.
* Delete any operator pod and services at top of caasapplicationprovisioner (v2) application worker to remove any previous pod-spec resources.
* Set remote state watcher's initial event IDs (when watcher is restarted). The remote state watcher is restarted when the charm is upgraded, and the pending pebble-ready event is dropped as a result. So carry over the pending event IDs to the new watcher instance.
* In worker/uniter/resolver, ensure we start a charm download if the charm dir doesn't exist (happens on upgrade).

## QA steps

```
# Download/build the sidecar version of the dex-auth charm
git clone https://github.com/canonical/dex-auth-operator.git
cd dex-auth-operator
git checkout sidecar
charmcraft build

# Bootstrap and install the pod-spec version
juju bootstrap microk8s
juju add-model test
juju deploy cs:dex-auth

# Finally, upgrade to the sidecar version
juju upgrade-charm dex-auth --path=./dex-auth.charm
```

### Steps to test storage carries over

First add the following to the charm's metadata.yaml and rebuild two versions, one on the main branch (podspec) and one on the sidecar branch (sidecar):

```
storage:
  testa:
    type: filesystem
    location: /testa
  testb:
    type: filesystem
    location: /testb
```

Then I used the following commands -- note writing something to both storages so we can ensure it's the same volume.

```
juju bootstrap microk8s
juju add-model t
juju model-config logging-config="<root>=DEBUG"
juju model-config logging-config="juju.kubernetes.provider.application=DEBUG" -m controller

juju deploy ../dex-auth_podspec_storage.charm --resource oci-image=quay.io/dexidp/dex:v2.28.1

microk8s.kubectl exec -n t dex-auth-0 -it -c dex-auth -- /bin/sh -c "echo AYE >/testa/aye"
microk8s.kubectl exec -n t dex-auth-0 -it -c dex-auth -- /bin/sh -c "echo BEE >/testb/bee"
microk8s.kubectl exec -n t dex-auth-0 -it -c dex-auth -- ls /testa
microk8s.kubectl exec -n t dex-auth-0 -it -c dex-auth -- ls /testb

juju upgrade-charm dex-auth --path=../dex-auth_sidecar_storage.charm

microk8s.kubectl exec -n t dex-auth-0 -it -c charm -- ls /testa
microk8s.kubectl exec -n t dex-auth-0 -it -c charm -- ls /testb
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1928778
